pkg/issuer/acme/issue.go

```diff
@@ -74,38 +78,88 @@ func (a *Acme) obtainCertificate(ctx context.Context, crt *v1alpha1.Certificate)
+	// state.
+	// The only way to obtain a certificate from an already 'valid' order is to
+	// call the orders GetCertificate function.
+	// You can see we do this below *iff* certSlice is nil.
+	if order.Status == acmeapi.StatusReady {
+		// generate a csr
+		template, err := pki.GenerateCSR(a.issuer, crt)
+		if err != nil {
+			// TODO: this should probably be classed as a permanant failure
+			return nil, nil, err
+		}
+
+		derBytes, err := pki.EncodeCSR(template, key)
+		if err != nil {
+			crt.UpdateStatusCondition(v1alpha1.CertificateConditionReady, v1alpha1.ConditionFalse, errorIssueError, fmt.Sprintf("Failed to generate certificate request: %v", err), false)
+			return nil, nil, err
+		}
+
+		// obtain a certificate from the acme server
+		certSlice, err = cl.FinalizeOrder(ctx, order.FinalizeURL, derBytes)
+		if err != nil {
+			// this handles an edge case where a certificate ends out with an order
+			// that is in an invalid state.
+			// ideally we would instead call GetCertificate on the ACME client
+			// instead of FinalizeOrder, which would save us creating a new order
+			// just to issue a new certificate.
+			// The underlying ACME client doesn't expose this though yet.
+			if acmeErr, ok := err.(*acmeapi.Error); ok {
+				if acmeErr.StatusCode >= 400 && acmeErr.StatusCode <= 499 {
+					crt.Status.ACMEStatus().Order.URL = ""
+				}
+			}
+			// TODO: should we also set the FailedValidation status
+			// condition here so back off can be applied?
+			crt.UpdateStatusCondition(v1alpha1.CertificateConditionReady, v1alpha1.ConditionFalse, errorIssueError, fmt.Sprintf("Failed to finalize order: %v", err), false)
+			a.Recorder.Eventf(crt, corev1.EventTypeWarning, errorIssueError, "Failed to finalize order: %v", err)
+			return nil, nil, fmt.Errorf("error getting certificate from acme server: %s", err)
+		}
+
+	}
+
+	// if the Certificate was marked 'valid', we need to retrieve the certificate
+	// from the URL specified on the Order resource.
+	if certSlice == nil {
+		certSlice, err = cl.GetCertificate(ctx, order.CertificateURL)
+		if err != nil {
+			crt.UpdateStatusCondition(v1alpha1.CertificateConditionReady, v1alpha1.ConditionFalse, errorIssueError, fmt.Sprintf("Failed to retrieve certificate: %v", err), false)
+			return nil, nil, fmt.Errorf("error retrieving certificate: %v", err)
+		}
+	}
+
+	if len(certSlice) == 0 {
+		return nil, nil, fmt.Errorf("invalid certificate returned from acme server")
 	}
 
-	derBytes, err := pki.EncodeCSR(template, key)
+	x509Cert, err := x509.ParseCertificate(certSlice[0])
 	if err != nil {
-		crt.UpdateStatusCondition(v1alpha1.CertificateConditionReady, v1alpha1.ConditionFalse, errorIssueError, fmt.Sprintf("Failed to generate certificate request: %v", err), false)
-		return nil, nil, err
+		return nil, nil, fmt.Errorf("failed to parse returned x509 certificate: %v", err.Error())
+	}
+
+	if a.Context.IssuerOptions.CertificateNeedsRenew(x509Cert) {
```

Comment:
Because we may have retrieved an existing certificate associated with an order, we must ensure that we have the correct private key for it and that the certificate is not nearing expiry (else we'd enter a loop of constantly requesting the same, expiring certificate).